### PR TITLE
Atomic design - Button inside input (WIP)

### PR DESCRIPTION
**Pull request by mistergone**
This pull request changes `.btn-inside-input` to atomic design (`.o-form__btn-inside-input`).

__Note:__ I labelled this "Work in Progress" because I have some concerns about padding and font-sizes in the input. I feel like the button as it stands is less differentiated than it was before because it has the same font size as the text and I don't know if the padding is doing the job. So, please discuss or correct me. 😄 

## Changes
- Moves code for __btn-inside-input modifier to forms.less
- Change __btn-inside-input modifier to use .a-text-input
- Change to padding-right to accomodate change in text size

## Testing
- I've created a `capital-framework-sandbox` branch for testing, [find it here](https://github.com/cfpb/capital-framework-sandbox/tree/btn-inside-input).

## Review
- @jimmynotjim @KimberlyMunoz @Scotchester (Please see my note above about the button differentiation)

## Screenshots
screen shot 2016-09-07 at 12 54 52 pm

## Notes
- This is currently a work in progress. I have not yet done testing in multiple browsers, for instance!

## Checklist

* [x] Changes are limited to a single goal (no scope creep)
* [x] Code can be automatically merged (no conflicts)
* [x] Code follows the standards laid out in the [front end playbook](https://github.com/cfpb/front-end)
* [x] Passes all existing automated tests
* [x] New functions include new tests
* [x] New functions are documented (with a description, list of inputs, and expected output)
* [x] Placeholder code is flagged
* [x] Visually tested in supported browsers and devices
* [ ] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)

### CODE REVIEW

**Comment by KimberlyMunoz** [comment]
I messed around with zoom to see if the padding-right is fragile or breaks under pressure there. It looks okay. The numbers you have are a little magical, but they work. The pattern will probably have to get revisited when we translate the "clear" button to other languages, but honestly I'm stumped as to how we could do it.

:+1: for now.

**Comment by cfarm** [comment]
I'm getting a less error with the sandbox so can't look at this in the browser, BUT i want to vote for the approach we discussed at the meeting today where the input and button are side by side like a default input/button pair, but the borders are styled to appear like they overlap.

**Comment by jimmynotjim** [comment]
> where the input and button are side by side like a default input/button pair, but the borders are styled to appear like they overlap.

Will this work if there's also a `submit` button after that? That's currently how this is most often used.

screen shot 2016-09-12 at 10 43 43 am

_ignore the broken styles with the icon to the left and the submit button height_

**Comment by mistergone** [comment]
@jimmynotjim - Thank you for pointing that out! I should be able to still make that work, I think. But now I'll make sure to test that.

**Comment by mistergone** [comment]
Entirely new code has been pushed! Highlights:

- Adds a new structure for btn-inside-input using containers
- Changes btn-inside-input to a molecule
- Changes display to `table` with `table-cell` elements, which allows the input portion to expand and the button to take up only the space necessary
- Adapts to localization! The text inside the "Clear" button can be altered as necessary - the button will take up the necessary space to display itself properly. Thus, other languages and other text in the button will work as expected.

TODO:
- This code doesn't properly handle focus, nor styles for modifiers to `a-text-input` such as `__error`, but those commits are close behind!

**Comment by mistergone** [comment]
Sandbox updated for new structure: https://github.com/cfpb/capital-framework-sandbox/tree/btn-inside-input

**Comment by mistergone** [comment]
__Also:__ Works in modern browsers, and works in IE9+ perfectly. Works acceptably in IE8 (putting a button next to it places the button on the next line, as with other inputs with buttons, so it seems like an acceptable fallback).

File: src/cf-forms/usage.md [hunk 1]
```diff
@@ -309,6 +309,7 @@ Provides sizeable margins between groups of form elements.
 </div>
 ```
 
+<<<<<<< HEAD
```
Comment (KimberlyMunoz) [inline src/cf-forms/usage.md]: Git conflict! ⚡

**Comment by mistergone** [comment]
futurama-good-news

The latest commits I pushed fix focus state, disabled state, and modifiers (`__error`, `__success`, `__warning`). Check out the sandbox repo I linked above for examples of these states and modifiers in action!

I haven't done extensive browser support testing yet, nor have I gone in and cleaned up the LESS to be perfect, but it works!

**Comment by mistergone** [comment]
__Sad news, everyone.__ My cross-browser testing has proven a disaster.

- __IE8__ - Doesn't look perfect, but is functional. Some borders appear offset, etc.
- __IE9, IE10__ - Spacing is off, some borders/outlines appear where they shouldn't
- __IE11__ - Spacing is off in a totally different way than IE9 and 10
- __Firefox__ - Spacing is off similar to IE9 and 10
- __Chrome__ - Works!

Once again, I feel like we're facing the question - is it worth more work to make this work, or do we scrap this and appeal to have the design changed?

**Comment by Scotchester** [comment]
I suppose I'd like to see screenshots of how badly each is failing.

**Comment by mistergone** [comment]
_Update_: Safari works!

Here's Firefox 48 inside saucelabs :

screen shot 2016-09-20 at 10 19 31 am

I'll post more screenshots of the other browsers soon.

**Comment by mistergone** [comment]
IE11 (in saucelabs), with its own unique failures:

screen shot 2016-09-20 at 10 28 19 am

**Comment by mistergone** [comment]
IE10 inside saucelabs, going its own way:

screen shot 2016-09-20 at 10 33 48 am

**Comment by mistergone** [comment]
After valiant efforts, I am giving up on fixing the "text other than Clear" issue and returning to the basic atomization of this molecule. Thanks for all your help and input @KimberlyMunoz and @jimmynotjim and @Scotchester - look for another commit that reverts to a simpler atomic version of this element.

**Comment by mistergone** [comment]
I've gone back to the original configuration of `btn-inside-input`, now `.m-btn-inside-input`. Sorry for all the insanity on this PR, but please review and comment. It's a much smaller diff now, so it should be easy to check over!

`capital-framework-sandbox` branch `btn-inside-input` has been updated to conform this version of the element, so please use it to check this out!

File: src/cf-forms/src/molecules/btn-inside-input.less [hunk 2]
```diff
@@ -0,0 +1,25 @@
+.m-btn-inside-input {
+    position: relative;
+
+    .a-text-input {
+        box-sizing: border-box;
+        width: 100%;
+        padding-right: unit( 75px / @btn-font-size, em );
+
+        &.input__super {
+            padding-right: unit( 90px / @super-btn-font-size, em );
```
Comment (jimmynotjim) [inline src/cf-forms/src/molecules/btn-inside-input.less]: Super input doesn't [exist any more](https://github.com/cfpb/capital-framework/pull/399).

File: src/cf-forms/src/molecules/btn-inside-input.less [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    }
+
+    .a-btn {
+        .u-link__no-border();
+        position: absolute;
+        right: unit( 15px / @btn-font-size, em );
+        top: 0;
+
+        /* pa11y fails unless the background color is explicitly set */
+        &__link.a-btn__secondary {
```
Comment (jimmynotjim) [inline src/cf-forms/src/molecules/btn-inside-input.less]: Is secondary currently being used in this situation? It's a color change only, and probably shouldn't be if it is.

File: src/cf-forms/usage.md [hunk 4]
```diff
@@ -209,27 +209,45 @@ See the 'Form icons' section below for guidance on adding icons to states.
...
-    <input type="text"
-           value="This is some really long text to make sure that the button doesn't overlap the content in such a way that this input becomes unusable."
-           title="Test input">
-    <button class="btn btn__link">
-        Clear
-        <span class="cf-icon cf-icon-delete"></span>
-    </button>
+
+<div class="m-btn-inside-input">
+    <div class="m-btn-inside-input_input-container">
```
Comment (jimmynotjim) [inline src/cf-forms/usage.md]: This container isn't used any more correct?

**Comment by mistergone** [comment]
Good catches, @jimmynotjim ! My reversion was a bit hasty. Fixed in the latest commit.

**Comment by jimmynotjim** [comment]
👍  thanks for following through on this, I know it was a pain.

**Comment by Scotchester** [comment]
:+1:

**Pull request merged**